File: cmd_youtube.py
# ملف: cmd_youtube.py
import os
import re
import subprocess
import tempfile
from telebot import types
from yt_dlp import YoutubeDL
import shutil

COMMANDS = ["يوت"]

YDL_OPTS = {
    "format": "bestaudio/best",
    "postprocessors": [{
        "key": "FFmpegExtractAudio",
        "preferredcodec": "mp3",
        "preferredquality": "192",
    }],
    "outtmpl": "",  # سيتم تحديده وقت التحميل
    "quiet": True,
    "noplaylist": True,
}

def search_youtube(query):
    safe_query = re.sub(r"[^\w\s]", "", query)
    return f"ytsearch1:{safe_query}"

def download_audio(query, chat_id):
    tmpdir = tempfile.mkdtemp()  # نستخدم mkdtemp بدلاً من with
    mp3_path = os.path.join(tmpdir, f"{chat_id}.mp3")
    YDL_OPTS["outtmpl"] = mp3_path
    try:
        with YoutubeDL(YDL_OPTS) as ydl:
            info = ydl.extract_info(search_youtube(query), download=True)
            title = info["entries"][0]["title"] if "entries" in info else info.get("title")
            thumbnail = info["entries"][0]["thumbnail"] if "entries" in info else info.get("thumbnail")
        return mp3_path, title, thumbnail, tmpdir
    except Exception as e:
        shutil.rmtree(tmpdir, ignore_errors=True)
        return None, None, None, None

def handle(bot, message, cmd_modules=None, game_modules=None, module_errors=None):
    text = message.text.strip()
    chat_id = message.chat.id

    if not text.lower().startswith("يوت "):
        return

    query = text[4:].strip()
    if not query:
        bot.reply_to(chat_id, "❌ اكتب اسم الأغنية بعد 'يوت'")
        return

    sent_msg = bot.send_message(chat_id, "👑 الإمبراطور يبحث ويجهّز MP3...")

    mp3_path, title, thumbnail, tmpdir = download_audio(query, chat_id)

    if not mp3_path:
        bot.edit_message_text("❌ حدث خطأ أثناء جلب الأغنية.", chat_id, sent_msg.message_id)
        return

    caption = f"🎵 {title}"

    # إرسال الغلاف أولاً
    if thumbnail:
        try:
            bot.send_photo(chat_id, thumbnail)
        except:
            pass

    # إرسال الصوت
    try:
        with open(mp3_path, "rb") as f:
            bot.send_audio(chat_id, f, title=title)
    except Exception as e:
        bot.send_message(chat_id, f"❌ حدث خطأ أثناء إرسال الصوت: {e}")

    # تنظيف الملفات بعد الإرسال
    shutil.rmtree(tmpdir, ignore_errors=True)
    bot.delete_message(chat_id, sent_msg.message_id)

File: cmd_youtube.js
// cmd_youtube.js
const { Telegraf } = require('telegraf');
const fs = require('fs');
const path = require('path');
const os = require('os');
const { execSync } = require('child_process');

async function handle(ctx) {
    const text = ctx.message?.text?.trim();
    if (!text || !text.toLowerCase().startsWith('يوت ')) return;

    const query = text.slice(4).trim();
    if (!query) return ctx.reply("❌ اكتب اسم الأغنية بعد 'يوت'");

    const chatId = ctx.chat.id;
    const sentMsg = await ctx.reply("👑 الإمبراطور يبحث ويجهّز الصوت...");

    // مجلد مؤقت للملفات
    const tempDir = fs.mkdtempSync(path.join(os.tmpdir(), 'yt-'));
    const outputTemplate = path.join(tempDir, '%(id)s.%(ext)s');

    try {
        // تحميل الصوت + الصورة باستخدام yt-dlp
        const cmd = `yt-dlp "ytsearch1:${query}" -x --audio-format opus --audio-quality 0 --output "${outputTemplate}" --write-thumbnail --no-playlist`;
        execSync(cmd, { stdio: 'ignore' });

        // البحث عن الملفات داخل tempDir
        const files = fs.readdirSync(tempDir);
        const audioFile = files.find(f => f.endsWith('.opus'));
        const thumbFile = files.find(f => f.endsWith('.jpg') || f.endsWith('.webp'));

        if (!audioFile) {
            return ctx.telegram.editMessageText(chatId, sentMsg.message_id, null, "❌ حدث خطأ أثناء تجهيز الصوت.");
        }

        const audioPath = path.join(tempDir, audioFile);
        const thumbPath = thumbFile ? path.join(tempDir, thumbFile) : null;

        const title = path.parse(audioFile).name;

        // إرسال الصوت كـ Voice مع الصورة واسم الأغنية
        if (thumbPath) {
            await ctx.telegram.sendVoice(chatId, { source: audioPath }, { caption: `🎵 ${title}`, thumb: { source: thumbPath } });
        } else {
            await ctx.telegram.sendVoice(chatId, { source: audioPath }, { caption: `🎵 ${title}` });
        }

        await ctx.telegram.deleteMessage(chatId, sentMsg.message_id);

    } catch (err) {
        await ctx.telegram.editMessageText(chatId, sentMsg.message_id, null, `❌ حدث خطأ: ${err.message}`);
    } finally {
        // حذف الملفات المؤقتة
        fs.rmSync(tempDir, { recursive: true, force: true });
    }
}

module.exports = { handle };
